docs/src/best-practices-js.md

```diff
@@ -0,0 +1,171 @@
+
+```bash
+npx playwright codegen
+```
+#### Use the VS Code extension to generate locators
+
+You can also use the [VS Code Extension](./getting-started-vscode.md) to generate locators as well as record a test. The VS Code extension also gives you a great developer experience when writing, running and debugging tests.
+
+## Use web first assertions
+
+Assertions are a way to verify that the expected result and the actual result matched or not. By using [web first assertions](./test-assertions.md) Playwright will wait until the expected condition is met. For Example when testing a toast message, if you click a button that makes a toast message appear you can test the toast message is there. If the toast takes half a second to appear Playwright will wait. Web assertions such as `toBeVisible()` will wait and retry.
+
+```js
+👍 await expect(page.getByText('welcome')).toBeVisible();
+```
+
+#### Don't use manual assertions
+
+Don't use manual assertions that are not awaiting the expect. In the code below the await is inside the expect rather than before it. When using assertions such as `isVisible()` the test wont wait a single second, it will just check the locator is there and return immediately. Use web first assertions such as `toBeVisible()` instead.
+
+```js
+👎 expect(await page.getByText('welcome').isVisible()).toBe(true);
+```
+
+Use web first assertions such as `toBeVisible()` instead.
+
+```js
+👍 await expect(page.getByText('welcome')).toBeVisible();
+```
+## Configure post-mortem debugging
+
+Use the playwright [trace viewer](./trace-viewer.md) for local debugging and post mortem debugging instead of videos and screenshots. The trace viewer gives you a full trace of your tests as a local PWA that can easily be shared. With the trace viewer you can view the timeline, inspect DOM snapshots for each action, view network requests and more.
+
+Traces are set to run on CI on the first retry of a failed test. However you can also run a trace locally when developing.
+
+```js
+npx playwright test --trace on
+```
+## Mock API responses
+
+Only test what you control. Don't try to test links to external sites or third party servers that you do not control. Not only is it time consuming and can slow down your tests but also you can not control the content of the page you are linking to, if there are cookie banners or overlay pages or anything else that might cause your test to fail.
+
+Intercepting the route with a mock response ensures the link is visible and clickable. Before hitting the link the route gets intercepted and a mock response is returned. Clicking the link results in a new page being opened containing the mock response rather than the actual page. We can then check this has the URL we expect.
```

Comment:
```suggestion
Here is an example of testing an external link.

Intercepting the route with a mock response ensures the link is visible and clickable. Before hitting the link the route gets intercepted and a mock response is returned. Clicking the link results in a new page being opened containing the mock response rather than the actual page. We can then check this has the URL we expect.
```